nexus/src/app/sagas/region_replacement_drive.rs

```diff
@@ -0,0 +1,1521 @@
+//! Today this is created either manually with omdb, or as a result of a
+//! physical disk being expunged. Affected VCRs are modified first by the region
+//! replacement start saga, which includes allocating a new replacement region.
+//! This then places the region replacement request into the state "Running".
+//! See that saga's documentation for more information.
+//!
+//! # why does the drive saga exist? #
+//!
+//! Region replacement is similar to instance migration in that it is initiated
+//! by Nexus but not directly controlled by it. Instance migration requires a
+//! source and destination Propolis to exist, and then Nexus waits for a
+//! callback to occur. For region replacement, it's Nexus' job to trigger
+//! either the Live Repair or Reconciliation operations via some Upstairs. Nexus
+//! then either receives a notification of success, or sees that the Volume is
+//! no longer in a degraded state as the result of some polling operation.
+//!
+//! Note: _it's very important that only_ the Upstairs can make the
+//! determination that a Volume is no longer degraded. Nexus should not be
+//! assuming anything. This is the _golden rule_ that this saga must follow.
+//!
+//! Volumes are put into a degraded state when the region replacement start saga
+//! modifies them: a blank region is swapped in one of the members of a region
+//! set. An Upstairs can still accept reads, writes, and flushes with only two
+//! out of three present in the set, but it's operating with a reduced
+//! redundancy.
+//!
+//! Volumes back higher level objects that users interact with: disks,
+//! snapshots, images, etc. Users can start and stop Upstairs by starting and
+//! stopping Instances. This interrupts any current operation on the Volume!
+//! This is ok: both operations were designed so that interruptions are not a
+//! problem, but it does stop progress.
+//!
+//! Say an Instance is running, and that Instance's propolis is performing a
+//! Live Repair. If a user stops that Instance, the propolis is torn down, and
+//! the Volume remains degraded. The next time that Volume is constructed and
+//! activated, the Upstairs will check each downstairs in the region set, see
+//! that there's a difference, and perform Reconciliation. If the user stops an
+//! Instance and does not start it again, that difference will remain.
+//!
+//! Nexus can at that point send the Volume to a Pantry and activate it, causing
+//! Reconciliation. At any time, the user can come along and start the Instance
+//! in question, which would take over the activation from the Pantry - this
+//! would cause that Reconciliation to fail, and the new propolis server would
+//! start its own Reconciliation. Again, the user may then stop the Instance,
+//! halting progress.
+//!
+//! This saga is responsible for driving forward th Volume repair process, by
+//! initiating repair operations. One invocation of this saga is not enough to
```

Comment:
You're right - if there were no interruptions, then yes it would be true. We would be able to join all three region replacement related sagas into one!

Reply:
Maybe change the comment to read "One invocation of this saga may not be enough to ..."
So, it's not suggested that we must run through multiple times, only that it is possible and or
likely that it will take more than one.

Reply:
I see now - changed the wording to "One invocation of this saga is most likely not enough" in 61dc32004